lib/aiken/rationals.ak

```diff
@@ -0,0 +1,747 @@
+  Rational { numerator: a_n * b_n, denominator: a_d * b_d }
+}
+
+test mul_1() {
+  ( expect_new(2, 3) |> mul(expect_new(3, 4)) ) == Rational {
+    numerator: 6,
+    denominator: 12,
+  }
+}
+
+test mul_2() {
+  ( expect_new(-2, 3) |> mul(expect_new(-3, 4)) ) == Rational {
+    numerator: 6,
+    denominator: 12,
+  }
+}
+
+test mul_3() {
+  (
+    expect_new(2, 5)
+      |> mul(expect_new(1, 8))
+      |> mul(expect_new(3, 10))
+      |> mul(expect_new(21, 100))
+      |> mul(expect_new(3, 5))
+      |> mul(expect_new(2, 8))
+      |> mul(expect_new(4, 10))
+      |> mul(expect_new(22, 100))
+      |> reduce
+  ) == Rational { numerator: 2079, denominator: 50000000 }
+}
+
+/// Division
+/// quotient of two rational values
+/// none if the second value is 0
+/// ```aiken
+/// div(new(x, y), new(a, b)) == new ( x*b, y*a )
+/// ```
+pub fn div(a: Rational, b: Rational) -> Option<Rational> {
+  recip(b) |> option.map(fn(x) { x |> mul(a) })
+}
+
+/// Expect division
+/// quotient of two rational values
+/// panics if the second value is 0 (expect)
+/// ```aiken
+/// expect_new(x, y) |> expect_div(expect_new(a, b)) == new(x*b, y*a)
+/// ```
+pub fn expect_div(a: Rational, b: Rational) -> Rational {
+  expect_recip(b) |> mul(a)
+}
+
+test div_1() {
+  ( expect_new(2, 3) |> expect_div(expect_new(3, 4)) ) == Rational {
+    numerator: 8,
+    denominator: 9,
+  }
+}
+
+test div_2() {
+  ( expect_new(2, 3) |> expect_div(expect_new(-3, 4)) ) == Rational {
+    numerator: -8,
+    denominator: 9,
+  }
+}
+
+/// Ratio of integers, a rational
+/// none if the second value is 0
+/// ```aiken
+/// new(x, 0) == None
+/// new(x, y) == Some(expect_new(x, y))
+/// ```
+pub fn new(numerator: Int, denominator: Int) -> Option<Rational> {
+  if denominator == 0 {
+    None
+  } else if denominator < 0 {
+    Some(Rational { numerator: -numerator, denominator: -denominator })
+  } else {
+    Some(Rational { numerator, denominator })
+  }
+}
+
+test new_zero() {
+  new(2, 0) == None
+}
+
+test new_1() {
+  new(2, 3) == Some(Rational { numerator: 2, denominator: 3 })
+}
+
+test new_2() {
+  new(-2, 3) == Some(Rational { numerator: -2, denominator: 3 })
+}
+
+test new_3() {
+  new(2, -3) == Some(Rational { numerator: -2, denominator: 3 })
+}
+
+test new_4() {
+  new(2, 4) == Some(Rational { numerator: 2, denominator: 4 })
+}
+
+test new_5() {
+  new(-2, -3) == Some(Rational { numerator: 2, denominator: 3 })
+}
+
+test new_6() {
+  new(-2, -4) == Some(Rational { numerator: 2, denominator: 4 })
+}
+
+/// Create a new Rational
+/// quotient of two integers
+/// panics if the second value is 0 (expect)
+/// ```aiken
+/// Some(expect_new(x, y)) == new(x, y)
+/// expect_new(x, 0) == error
+/// ```
+pub fn expect_new(numerator: Int, denominator: Int) -> Rational {
+  if denominator == 0 {
+    error @"Division by 0"
+  } else if denominator < 0 {
+    Rational { numerator: -numerator, denominator: -denominator }
+  } else {
+    Rational { numerator, denominator }
+  }
+}
```

Comment:
Now looking at the implementation of both functions, I don't really see the benefit of this function compared to:

```
expect Some(r) = rational.new(14, 42)
```

Reply:
I don't like using `expect`. For my use cases, I leave the traces in the final script because my designs are meant to be used by anyone: directly or through a front-end. The error messages are therefore not meant for me. They are meant to guide the user/front-end to the proper usage. Since the `expect` error messages are only really useful to smart contract developers, I try to avoid using it. The error message that is thrown with `expect_new` is much more user friendly. While documentation can help close the gap, the error message in `expect_new` is a better starting point IMO. If I could do what is suggested [here](https://github.com/aiken-lang/aiken/issues/489#issuecomment-1584305945), I would reconsider.

Reply:
@fallen-icarus in that regard, I remember someone suggesting that we allow for custom error message on expect's

I forget if it was @waalge or @KristianBalaj ? if this was one of you do you remember where you originally posted about that?

Reply:
Yup, I've posted it in this issue in the language repository https://github.com/aiken-lang/aiken/issues/489#issuecomment-1584305945

Reply:
Ah great, I couldn't find it for some reason earlier.

Reply:
https://github.com/aiken-lang/stdlib/pull/47#discussion_r1232154125 -> That's actually an argument I can relate to. And I'd argue that returning `Option` is then going in the direction of self-documenting code. With the addition of #489 I feel like we have a nice compromise to make the use of `expect` at call-site even more user-friendly? 

I'd rather solve that in a more holistic way by providing better language features overall than crippling the standard library with duplicate functions and weird prefixes 😬

Reply:
@KtorZ I agree